uamqp/session.py

```diff
@@ -214,13 +215,61 @@ def _outgoing_transfer(self, delivery):
         if self.remote_incoming_window <= 0:
             delivery.transfer_state = SessionTransferState.Busy
         else:
+            payload = delivery.frame['payload']
+            payload_size = len(payload)
+
             delivery.frame['delivery_id'] = self.next_outgoing_id
-            self._connection._process_outgoing_frame(self.channel, TransferFrame(**delivery.frame))
+            # calculate the transfer frame encoding size excluding the payload
+            delivery.frame['payload'] = b""
+            encoded_frame = encode_frame(TransferFrame(**delivery.frame))[1]
+            transfer_overhead_size = len(encoded_frame)
+
+            # available size for payload per frame is calculated as following:
+            # remote max frame size - transfer overhead (calculated) - header (8 bytes)
+            available_frame_size = self._connection._remote_max_frame_size - transfer_overhead_size - 8
+
+            start_idx = 0
+            remaining_payload_cnt = payload_size
+            # encode n-1 frames if payload_size > available_frame_size
+            while remaining_payload_cnt > available_frame_size:
+                tmp_delivery_frame = {
+                    'handle': delivery.frame['handle'],
+                    'delivery_tag': delivery.frame['delivery_tag'],
+                    'message_format': delivery.frame['message_format'],
+                    'settled': delivery.frame['settled'],
+                    'more': True,
+                    'rcv_settle_mode': delivery.frame['rcv_settle_mode'],
+                    'state': delivery.frame['state'],
+                    'resume': delivery.frame['resume'],
+                    'aborted': delivery.frame['aborted'],
+                    'batchable': delivery.frame['batchable'],
+                    'payload': payload[start_idx:start_idx+available_frame_size],
+                    'delivery_id': self.next_outgoing_id
```

Comment:
next_outgoing_id is increased in the end after all the transfer frames are being sent, this is kinda related to your "Don't all these values need modifying if the number of frames sent > 1?" question.

from my understanding, it's only needed to be increased after all the transfer frames of a message have been sent out.

Reply:
next_outgoing_id is increased in the end after all the transfer frames are being sent, this is kinda related to your "Don't all these values need modifying if the number of frames sent > 1?" question.

from my understanding, it's only needed to be increased after all the transfer frames of a message have been sent out.

Reply:
Okay - so it sounds like the reason we don't need to increment the other values is because all these transfers are sent under the same delivery ID.... so it's a single frame, using only the `more` field to indicate multiple payloads.